### DISCUSSION TITLE
Orangepi5 Rockchip RK3588S hardware decoder through video4linux virtual device

### DISCUSSION THREAD

**Discussion by eusoubrasileiro** [Motion-Project/motion#1710] [Q&A] [answered]
Hello everyone I just bought my Orangepi5. 

I was testing this [custom build of ffmpeg](https://github.com/hbiyik/FFmpeg) that uses the MPP from rockchip.  

It properlly works outside motion I tested with `ffmpeg -stats -rtsp_transport tcp -y -c:v  h264_rkmpp_decoder -i rtsp://admin:gig1684A@ipcam-right-aisle:554/ch0_1.264 -f null -` etc... 
But for motion I could not make `hevc_rkmpp_decoder` or `h264_rkmpp_decoder` to work. 
I have been playing with motion for 4 years I guess so configuration isn't a isse.
The problem that I see is that they don't suppor 'D' direct rendering 1 method. 

```
ffmpeg -decoders | grep rkmpp
ffmpeg version e243e8d Copyright (c) 2000-2023 the FFmpeg developers
  built with gcc 11 (Ubuntu 11.4.0-1ubuntu1~22.04)
  configuration: --disable-ffprobe --disable-ffplay --disable-outdevs --disable-indevs --disable-devices --enable-shared --prefix=/usr/local --enable-rkmpp --enable-version3 --enable-libdrm
  libavutil      58.  2.100 / 58.  2.100
  libavcodec     60.  3.100 / 60.  3.100
  libavformat    60.  3.100 / 60.  3.100
  libavdevice    60.  1.100 / 60.  1.100
  libavfilter     9.  3.100 /  9.  3.100
  libswscale      7.  1.100 /  7.  1.100
  libswresample   4. 10.100 /  4. 10.100
 V..... av1_rkmpp_decoder    av1 (rkmpp decoder ) (codec av1)
 V..... h263_rkmpp_decoder   h263 (rkmpp decoder ) (codec h263)
 V..... h264_rkmpp_decoder   h264 (rkmpp decoder ) (codec h264)
 V..... hevc_rkmpp_decoder   hevc (rkmpp decoder ) (codec hevc)
 V..... mpeg1_rkmpp_decoder  mpeg1 (rkmpp decoder ) (codec mpeg1video)
 V..... mpeg2_rkmpp_decoder  mpeg2 (rkmpp decoder ) (codec mpeg2video)
 V..... mpeg4_rkmpp_decoder  mpeg4 (rkmpp decoder ) (codec mpeg4)
 V..... vp8_rkmpp_decoder    vp8 (rkmpp decoder ) (codec vp8)
 V..... vp9_rkmpp_decoder    vp9 (rkmpp decoder ) (codec vp9)
```

I was just wondering If creating a virtual device with `video4linux` using ffmpeg and that specific decoder would make it usable?

Does anybody tried this before?

**Comment by eusoubrasileiro** [comment, marked as answer]
This ffmpeg port can't work with this decoder hevc_rkmpp_decoder etc due missing direct rendering or whatever other reason.

### But it works 

Hardware decoding is WORKING (somehow).
I'm almost sure it's working properly with hardware decoding **when we select 'hevc' or 'h264' as decoder type** in motion *.confs! 
They were supposedly to be software decoding but... 
Somehow Rockchip Mpp is automatically offloading the decoding tasks to the hardware.
With 3 cameras and 6 rtsp streams it's using only 4% of cpu. I will confirm with more cameras.

For future I will try this new version of Rockchip ffmpeg port [this here ](https://github.com/nyanmisaka/ffmpeg-rockchip)

**Reply by vukitoso** [reply]
> With 3 cameras and 6 rtsp streams it's using only 4% of cpu.

I have Radxa Rock-3a (RK3568).

@eusoubrasileiro, I build from the sources https://github.com/nyanmisaka/ffMpeg-rockchip.
Do I need to build Motion with the parameter "--with-ffmpeg="?
